src/mcts/node.cc

```diff
@@ -197,35 +249,51 @@ void Node::CreateEdges(const MoveList& moves) {
 Node::ConstIterator Node::Edges() const { return {edges_, &child_}; }
 Node::Iterator Node::Edges() { return {edges_, &child_}; }
 
+// Recalculate n_ from real children visits.
+// This is needed if a node was proved to be certain in a prior
+// search and later gets to be root of search.
+void Node::ReComputeN() {
+  if (n_ > 1) {
+    uint32_t visits = 1;
+    for (const auto& child : Edges()) visits += child.GetN();
+    n_ = visits;
+    }
+  n_in_flight_ = 0;
+}
+
+
 float Node::GetVisitedPolicy() const { return visited_policy_; }
 
+float Node::GetQ() const {
+  if (parent_) {
```

Comment:
I don't really like relying much on a parent, as it was a temporary optimization only.
Is Q conceptually edge's or node's property?
If edge's:  Why do we need to check "if edge is certain" in a node rather than edge?
If node's: Why do we store that it's certain in an edge rather than a node?

Node should not go to parent to get some information about itself.

Reply:
I implemented it as to be as minimal invasive to the current codebase. You are correct that overall it would be "nicer" to move from a node centric style to a edge.node or even edge centric style. I left the parts of the code as is, if they checked for terminal/certain status by node. I would propose changing this only after CP is tested and with the complete codebase and future changes in mind. 

So while I feel that there is currently a shift towards a more edge centric view with your new detachment scheme which might store N in the edge as well, and also even the original A0 seems rather edge centric), I would still make that a seperate PR sometime later.